File: python/connection.py
import bitarray
import message
import random
import threading


class Connection(threading.Thread):

    def __init__(self, sock, initiate_handshake, peer):
        super(Connection, self).__init__()
        self.sock = sock
        self.initiate_handshake = initiate_handshake
        self.peer = peer

        self.other_peer_id = -1
        self.other_arr = bitarray.bitarray(peer.num_pieces)
        self.other_arr.setall(False)
        self.interested = False
        self.choked = True
        self.pieces_received = 0

    def run(self):
        try:
            if self.initiate_handshake:
                self.send(message.handshake(self.peer.peer_id))

            while self.peer.missing_pieces() or self.peer.neighbor_missing_pieces():
                self.respond(self.rcv())

            self.sock.close()
            self.peer.connections.remove(self)
            print(f'Peer {self.peer.peer_id} closed connection with Peer {self.other_peer_id}')
        except ConnectionError:
            print(f'Peer {self.peer.peer_id}-{self.other_peer_id} connection was closed')

    def respond(self, msg):
        if msg.type == message.HANDSHAKE:
            self.other_peer_id = int.from_bytes(msg.payload, 'big')
            if not self.initiate_handshake:
                self.send(message.handshake(self.peer.peer_id))
            if self.peer.has_file:
                self.send(message.bitfield(self.peer.arr))
        elif msg.type == message.CHOKE:
            pass
        elif msg.type == message.UNCHOKE:
            self._request_piece()
        elif msg.type == message.INTERESTED:
            self.interested = True
        elif msg.type == message.NOT_INTERESTED:
            self.interested = False
        elif msg.type == message.HAVE:
            index = int.from_bytes(msg.payload, 'big')
            self.other_arr[index] = True
            self._notify_interest()
        elif msg.type == message.BITFIELD:
            self.other_arr.clear()
            self.other_arr.frombytes(msg.payload)
            self._notify_interest()
        elif msg.type == message.REQUEST:
            index = int.from_bytes(msg.payload, 'big')
            self.send(message.piece(index, self.peer.pieces[index]))
        elif msg.type == message.PIECE:
            index = int.from_bytes(msg.payload[:4], 'big')
            piece = msg.payload[4:]
            self.peer.pieces[index] = piece
            with open(self.peer.filename, 'wb') as f:
                f.seek(index * self.peer.piece_size)
                f.write(piece)
            self.peer.arr[index] = True
            self.pieces_received += 1
            for conn in self.peer.connections:
                conn.send(message.have(index))
            self._request_piece()

    def _request_piece(self):
        diff = self.other_arr & ~self.peer.arr
        start_pos = diff.search(bitarray.bitarray('1'))
        if len(start_pos) > 0:
            self.send(message.request(random.choice(start_pos)))

    def _notify_interest(self):
        diff = self.other_arr & ~self.peer.arr
        if diff.count() > 0:
            self.send(message.interested())
        else:
            self.send(message.not_interested())

    def rcv(self):
        raw = b''
        while len(raw) < 4:
            raw += self.sock.recv(4)
        length = 28 if raw == b'P2PF' else int.from_bytes(raw, 'big')
        while len(raw) < 4 + length:
            raw += self.sock.recv(length)
        msg = message.parse(raw)
        print(f'Peer {self.peer.peer_id} received {msg.get_type()} from Peer {self.other_peer_id}')
        return msg

    def send(self, msg):
        self.sock.send(msg)
        print(f'Peer {self.peer.peer_id} sent {message.parse(msg).get_type()} to Peer {self.other_peer_id}')

File: java2/Connection.java
import java.io.EOFException;
import java.io.File;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.RandomAccessFile;
import java.net.Socket;
import java.net.SocketException;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.BitSet;
import java.util.concurrent.ThreadLocalRandom;

public class Connection implements Runnable {
    public BitSet otherBitarray;
    public int otherPeerId = -1;
    public boolean interested = false;
    public boolean choked = true;
    public int piecesReceived = 0;

    private final boolean initiateHandshake;
    private final Socket socket;
    private final Peer peer;

    private ObjectInputStream in;
    private ObjectOutputStream out;
    private RandomAccessFile raf;

    public Connection(boolean initiateHandshake, Socket socket, Peer peer) {
        this.initiateHandshake = initiateHandshake;
        this.socket = socket;
        this.peer = peer;
        otherBitarray = new BitSet(peer.numPieces);
        otherBitarray.clear();

        if (peer.hasFile == 0) {
            try {
                String path = String.format("./peer_%s/%s", peer.peerId, peer.filename);
                File file = new File(path);
                file.getParentFile().mkdirs();
                file.createNewFile();
                raf = new RandomAccessFile(path, "rw");
            } catch (IOException e) {
                System.err.println(e.getMessage());
            }
        }
    }

    public void run() {
        try {
            out = new ObjectOutputStream(socket.getOutputStream());
            in = new ObjectInputStream(socket.getInputStream());
            if (initiateHandshake) {
                send(Message.handshake(peer.peerId));
            }

            while (peer.missingPieces() || peer.neighborsMissingPieces()) {
                respond(rcv());
            }

            in.close();
            out.close();
            socket.close();
            System.out.printf("Peer %s closed connection with Peer %s%n", peer.peerId, otherPeerId);
        } catch (SocketException | EOFException e) {
            System.out.printf("Peer %s-%s connection was closed%n", peer.peerId, otherPeerId);
        } catch (IOException | ClassNotFoundException e) {
            System.err.println(e.toString());
        }
        peer.connections.remove(this);
    }

    private Message rcv() throws IOException, ClassNotFoundException {
        byte[] msg = (byte[]) in.readObject();
        System.out.printf("Peer %s received %s from Peer %s%n", peer.peerId, getType(msg), otherPeerId);
        return Message.parse(msg);
    }

    public void send(byte[] msg) throws IOException {
        out.writeObject(msg);
        System.out.printf("Peer %s sent %s to Peer %s%n", peer.peerId, getType(msg), otherPeerId);
    }

    private String getType(byte[] msg) {
        return Message.parse(msg).type.toString().toLowerCase();
    }

    private void respond(Message msg) throws IOException {
        switch (msg.type) {
            case HANDSHAKE -> {
                otherPeerId = byteArrayToInt(msg.payload);
                if (!initiateHandshake) {
                    send(Message.handshake(peer.peerId));
                }
                synchronized (peer.bitarray) {
                    if (!peer.bitarray.isEmpty()) {
                        send(Message.bitfield(peer.bitarray));
                    }
                }
            }
            case CHOKE -> {}
            case UNCHOKE -> requestPiece();
            case INTERESTED -> interested = true;
            case NOT_INTERESTED -> interested = false;
            case HAVE -> {
                int index = byteArrayToInt(msg.payload);
                otherBitarray.set(index, true);
                notifyInterest();
            }
            case BITFIELD -> {
                otherBitarray = BitSet.valueOf(msg.payload);
                notifyInterest();
            }
            case REQUEST -> {
                int index = byteArrayToInt(msg.payload);
                send(Message.piece(index, peer.pieces.get(index)));
            }
            case PIECE -> {
                int index = byteArrayToInt(Arrays.copyOfRange(msg.payload, 0, 4));
                byte[] piece = Arrays.copyOfRange(msg.payload, 4, msg.payload.length);
                peer.pieces.put(index, piece);
                raf.seek(index * peer.pieceSize);
                raf.write(piece);
                synchronized (peer.bitarray) {
                    peer.bitarray.set(index);
                }
                piecesReceived++;
                for (Connection conn : peer.connections) {
                    conn.send(Message.have(index));
                }
                requestPiece();
            }
        }
    }

    private int byteArrayToInt(byte[] bytes) {
        return ByteBuffer.wrap(bytes).getInt();
    }

    private void requestPiece() throws IOException {
        BitSet diff = missingBitarray();
        int index = -1;
        if (!diff.isEmpty()) {
            while (index < 0 || index > peer.numPieces - 1) {
                index = ThreadLocalRandom.current().nextInt(0, peer.numPieces - 1);
                index = diff.nextSetBit(index);
            }
            send(Message.request(index));
        }
    }

    private void notifyInterest() throws IOException {
        BitSet diff = missingBitarray();
        if (diff.isEmpty()) {
            send(Message.notinterested());
        } else {
            send(Message.interested());
        }
    }

    private BitSet missingBitarray() {
        synchronized (peer.bitarray) {
            BitSet diff = (BitSet) otherBitarray.clone();
            diff.andNot(peer.bitarray);
            return diff;
        }
    }
}
